### PR TITLE
Import Nightbot Custom Commands & Timers

### PR DESCRIPTION
**Pull request by redbeardymcgee**
This is a quick and dirty first pass at fetching a [user's custom commands](https://api-docs.nightbot.tv/#get-custom-commands) from their Nightbot.

I ran `prettier` over the file once to reduce some of the line lengths so I could read the file better. There were a couple of `eslint` warnings that I manually fixed up. Apologies for the whitespace churn in this initial commit!

Meaningful changes begin at L#188, but a lot of this will probably be rewritten after the Nightbot OAuth token includes the `commands` scope. I think that happens [here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3) but I'm pretty sure I can't test a change to that locally since the oauth callbacks won't work on my localhost.

### CODE REVIEW

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 1]
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+            onClick={importPlaylist}
+          >
+            Import playlist
+          </Button>
+          <Button
+            color="primary"
+            variant="contained"
+            disabled={user === 'Not Authorized'}
+            onClick={() => {
+              console.log(fetchCustomCommands());
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: you don't need it here as you are using it inside importCustomCommands

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 2]
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+            disabled={user === 'Not Authorized'}
+            onClick={importPlaylist}
+          >
+            Import playlist
+          </Button>
+          <Button
+            color="primary"
+            variant="contained"
+            disabled={user === 'Not Authorized'}
+            onClick={() => {
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
            onClick={importCustomCommands}
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx (hunk 1 above)
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 3]
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+              console.log(fetchCustomCommands());
+              return importCustomCommands;
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
```

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 4]
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+            }}
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]:
```suggestion
```

**Comment by redbeardymcgee** [comment]
At this point, I have a working version of the full custom command import. Support for special nightbot features (if any such features exist) is not implemented.

However, my local instance of the dashboard returns a 404 when deleting commands. This includes the imported commands, or commands manually created through the UI at `/commands/customcommands`. Works fine with the same instance of my local bot through `dash.sogebot.xyz`, so my local environment has to be broken.

**Comment by redbeardymcgee** [comment]
This is, in some sense of the word, complete. However, there are some weird things about Nightbot that I think are missing direct analogues in Sogebot.

First, Nightbot allows commands with no prefix, while Sogebot requires exactly the `!` character to begin any command. It's a fun feature sometimes though. For example, you could set up non-prefixed commands to make the bot help build a `Kappa` pyramid (or to interrupt some emote pyramid for the opposite joke lol). For now, I'm allowing this to fail but I could probably just prefix it with an exclamation point.

Second, I didn't figure out the meaning of the `filter` key on responses. Probably an empty string is fine, but I'd like some confirmation.

Third, Nightbot supports a variety of [special variables](https://docs.nightbot.tv/commands/variableslist) within its commands. [Some of them are not worth porting or converting](https://docs.nightbot.tv/variables/djkhaled), but at least [a few of them](https://docs.nightbot.tv/variables/channel) probably have direct analogues in Sogebot. I'm not sure which ones should be converted to Sogebot's conventions, if any.

Fourth, and I believe *finally*, this feature requires a new scope `commands`. I noted in the original comment on this PR that I think I found where to add the scope, but I can't test it locally to confirm. When I did try to add the scopes [here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3), the query parameters generated when clicking `Authorize` still only had `scope=song_requests_playlist` without adding `commands`. I guess I didn't make the change correctly, or that's not the right place after all :)

Thanks again for all your patience while I work through this, and for building this cool bot to make it possible for me to do this :)

**Comment by sogehige** [comment]
> First, Nightbot allows commands with no prefix, while Sogebot requires exactly the ! character to begin any command. It's a fun feature sometimes though. For example, you could set up non-prefixed commands to make the bot help build a Kappa pyramid (or to interrupt some emote pyramid for the opposite joke lol). For now, I'm allowing this to fail but I could probably just prefix it with an exclamation point.

You can probably use keyword instead of command

> Second, I didn't figure out the meaning of the filter key on responses. Probably an empty string is fine, but I'd like some confirmation.

That's ok

> Third, Nightbot supports a variety of [special variables](https://docs.nightbot.tv/commands/variableslist) within its commands. [Some of them are not worth porting or converting](https://docs.nightbot.tv/variables/djkhaled), but at least [a few of them](https://docs.nightbot.tv/variables/channel) probably have direct analogues in Sogebot. I'm not sure which ones should be converted to Sogebot's conventions, if any.

Hmm, porting some of those may be tricky, although its up to you if you want to do some porting. I would rather add some warning or notification that user should revisit imported commands and adjust them manually for bot.

> Fourth, and I believe finally, this feature requires a new scope commands. I noted in the original comment on this PR that I think I found w[here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3) to add the scope, but I can't test it locally to confirm. When I did try to add the scopes here, the query parameters generated when clicking Authorize still only had scope=song_requests_playlist without adding commands. I guess I didn't make the change correctly, or that's not the right place after all :)

That's correct, you need to update it there, then I need to rebuild service and deploy it on server.

**Comment by redbeardymcgee** [comment]
I don't really think the effort of trying to port Nightbot's special variables is worthwhile until we have a user who depends on it. I just added a snackbar notif for now to go look at the imported commands.

I don't think I have any other notes for this PR then. If no one else has any objections to the code, I'm ready to call this done.

**Comment by sogehige** [comment]
Added "commands" scope to service, which should be deployed in a while so you can test it

**Comment by redbeardymcgee** [comment]
Thanks, the `commands` scope worked but I forgot to add the `timers` scope as well. After that is included, this will be feature complete.

IMO, playlist, custom commands and timers are the main 3 features that Nightbot users actually use a lot. Someone may, someday, request another scope to import but that can be built up in like 30 minutes now that I made 3 examples. 

Ready to merge whenever you say :)

File: ui.admin-mui/src/components/Settings/Import/nightbot.tsx [hunk 5]
```diff
@@ -45,152 +51,90 @@ const PageSettingsModulesImportNightbot: React.FC<{
...
+          </Button>
+          <Button
+            color="primary"
+            variant="contained"
+            disabled={user === 'Not Authorized'}
+            onClick={() => importTimers(accessToken)}
+          >
+            Import timers
+          </Button>
         </Stack>
```
Comment (sogehige) [inline ui.admin-mui/src/components/Settings/Import/nightbot.tsx]: Thinking of little bit better styling, wdyt?

```suggestion
       <Stack direction="row" spacing={1} sx={{ pt: 1 }}>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importPlaylist(accessToken)}
          >
            Import playlist
          </Button>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importCustomCommands(accessToken)}
          >
            Import commands
          </Button>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importTimers(accessToken)}
          >
            Import timers
          </Button>
        </Stack>
```

**Comment by sogehige** [comment]
Can you comment on my suggestion? Then we can merge :)

**Comment by sogehige** [comment]
Timers 

> Thanks, the `commands` scope worked but I forgot to add the `timers` scope as well. After that is included, this will be feature complete.
> 
> IMO, playlist, custom commands and timers are the main 3 features that Nightbot users actually use a lot. Someone may, someday, request another scope to import but that can be built up in like 30 minutes now that I made 3 examples.
> 
> Ready to merge whenever you say :)

`timers` scope added to service

**Pull request merged**